### PR TITLE
*: Prevent Leader from accidentally hanging up data inconsistency

### LINKED ISSUE #598: [bug] 🐛 force restart leader may cause data inconsistency
**Describe the problem**

**To Reproduce**

1. sysbench keep writing data then kill leader pod or power down the leader node

**Expected behavior**

**Environment:**

- RadonDB MySQL version:

### PR DESCRIPTION
**Pull request by zhl003**
### What type of PR is this?

/enhancement
### Which issue(s) this PR fixes?

Fixes #598

### What this PR does?

Summary:
```
1. Add leader postStart,preStop cmd in xenon container   
2. Add leaderstop cmd in xenon conf
3. Adjust project structure
```

### CODE REVIEW

**Review by acekingke — APPROVED** [review]
LGTM

**Review by qianfen2021 — APPROVED** [review]
LGTM

File: cmd/xenon/main.go [hunk 1]
```diff
@@ -1,20 +1,83 @@
...
+	"io/ioutil"
 	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
 
+	"github.com/bitly/go-simplejson"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/jmoiron/sqlx"
```
Comment (acekingke) [inline cmd/xenon/main.go:18]: _ "github.com/go-sql-driver/mysql"  is ok， Do not use "github.com/jmoiron/sqlx"

File: cmd/xenon/main.go [hunk 2]
```diff
@@ -54,7 +117,86 @@ func leaderStart() error {
 }
 
 func leaderStop() error {
-	return PatchRoleLabelTo(myself(string(Follower)))
+	log.Infof("leader stop started")
+	// first load any configuration provided (e.g. via environment variables)
+	cfg := loadConfiguration()
+	dsn := mysqlUser + ":" + cfg.password + "@tcp(" + mysqlHost + ")/"
+	server := MySQLServer{DSN: dsn}
+	Conn, err := server.GetNewDBConn()
```
Comment (acekingke) [inline cmd/xenon/main.go:125]: `conn` must close when finished

File: cmd/xenon/main.go [hunk 3]
```diff
@@ -54,7 +117,86 @@ func leaderStart() error {
...
+	if err != nil {
+		log.Error(err)
+	}
+	IsLeader := IsLeader(Conn)
```
Comment (acekingke) [inline cmd/xenon/main.go:129]: when error,  to here is will crash.

File: cmd/xenon/main.go [hunk 4]
```diff
@@ -54,7 +117,86 @@ func leaderStart() error {
...
+	if !IsLeader {
+		log.Info("I am not the leader")
+		os.Exit(0)
+	}
+	// Step 1: disable event scheduler
+	log.Info("Disabling event scheduler")
+	stmt, err := SetEventScheduler(Conn, false)
```
Comment (acekingke) [inline cmd/xenon/main.go:136]: the same question.

File: cmd/xenon/main.go [hunk 5]
```diff
@@ -54,7 +117,86 @@ func leaderStart() error {
...
+		log.Error(err)
+	}
+	log.Infof("set event scheduler to false: %s", stmt)
+	// Step 2: set readonly
+	log.Info("Setting readonly")
+	stmt, err = SetReadOnly(Conn, true)
+	if err != nil {
+		log.Error(err)
+	}
+	log.Infof("set readonly to true: %s", stmt)
```
Comment (acekingke) [inline cmd/xenon/main.go:147]: the same, when error , it will crash.

File: cmd/xenon/main.go [hunk 6]
```diff
@@ -54,7 +117,86 @@ func leaderStart() error {
...
+	// Step 3: TODO: check long running writes
+	log.Info("Checking long running writes")
+	num, stmt, err := CheckLongRunningWrites(Conn, 4)
+	if err != nil {
+		log.Error(err)
+	}
+	log.Infof("%d,long running writes: %s", num, stmt)
```
Comment (acekingke) [inline cmd/xenon/main.go:154]: the same, when error , it will crash.

File: cmd/xenon/main.go [hunk 7]
```diff
@@ -54,7 +117,86 @@ func leaderStart() error {
...
+	// 	log.Error(err)
+	// }
+	// log.Info("set max connections to 1: ", stmt)
+	// Step 5: kill threads
+	log.Info("Killing threads")
+	stmt, err = KillThreads(Conn)
+	if err != nil {
+		log.Error(err)
+	}
+	log.Info("killed threads: ", stmt)
```
Comment (acekingke) [inline cmd/xenon/main.go:168]: the same, when error , it will crash.

File: cmd/xenon/main.go [hunk 8]
```diff
@@ -54,7 +117,86 @@ func leaderStart() error {
...
+	// Step 6: FlushTablesWithReadLock
+	log.Info("Flushing tables with read lock")
+	stmt, err = FlushTablesWithReadLock(Conn)
+	if err != nil {
+		log.Error(err)
+	}
+	log.Info("flushed tables with read lock: ", stmt)
```
Comment (acekingke) [inline cmd/xenon/main.go:175]: the same, when error , it will crash.

File: cmd/xenon/main.go [hunk 9]
```diff
@@ -69,13 +211,71 @@ func readiness() error {
...
+					log.Infof("I am the leader, skip GTID_SUBSET")
+					return nil
+				}
+				mySet = gtid.ExecutedGTIDSet
+			}
+			if gtid.Raft == string(Leader) {
+				leaderSet = gtid.ExecutedGTIDSet
+			}
+		}
+		db, err := server.GetNewDBConn()
```
Comment (acekingke) [inline cmd/xenon/main.go:247]: db need close.

File: cmd/xenon/main.go [hunk 10]
```diff
@@ -86,3 +286,366 @@ func myself(role string) MySQLNode {
...
+	cfg.podName = os.Getenv("PODNAME")
+	if cfg.podName == "" {
+		cfg.podName, _ = os.Hostname()
+	}
+	log.Debugf("PODNAME set to: %s", cfg.podName)
+	xenonConfFile, err := os.Open(xenonConf)
+	if err != nil {
+		log.Fatal(err)
+	}
+	contents, err2 := ioutil.ReadAll(xenonConfFile)
```
Comment (acekingke) [inline cmd/xenon/main.go:387]: `defer xenonConfFile.Close()` should at here.

File: cmd/xenon/main.go [hunk 11]
```diff
@@ -86,3 +286,366 @@ func myself(role string) MySQLNode {
...
+	ticker := time.NewTicker(time.Second * 1)
+	defer ticker.Stop()
+	timeoutExceeded := time.After(timeout)
+	for {
+		select {
+		case <-timeoutExceeded:
+			return false, fmt.Errorf("db connection failed after %s timeout", timeout)
+		case <-ticker.C:
+			_, err := sqlx.Connect("mysql", server.DSN)
+			if err == nil {
```
Comment (acekingke) [inline cmd/xenon/main.go:613]: when success， need close.

File: config/samples/mysql_v1alpha1_backup.yaml [hunk 12]
```diff
@@ -8,4 +8,4 @@ spec:
   # hostname if empty, use the leader as hostname
   hostName: sample-mysql-0
   clusterName: sample
-  # nfsServerAddress: ""
+  # nfsServerAddress: ""
```
Comment (acekingke) [inline config/samples/mysql_v1alpha1_backup.yaml:11]: new lines.

File: hack/xenon/Dockerfile [hunk 13]
```diff
@@ -37,4 +37,4 @@ EXPOSE 8801
 VOLUME ["/var/lib/xenon", "/etc/xenon"]
 
 ENTRYPOINT ["xenon"]
-CMD ["-c", "/etc/xenon/xenon.json"]
+CMD ["-c", "/etc/xenon/xenon.json"]
```
Comment (acekingke) [inline hack/xenon/Dockerfile:40]: new lines.

File: cmd/xenon/main.go [hunk 14]
```diff
@@ -86,3 +286,366 @@ func myself(role string) MySQLNode {
...
+}
+
+func (server *MySQLServer) WaitforMySQL(timeout time.Duration) (bool, error) {
+	ticker := time.NewTicker(time.Second * 1)
+	defer ticker.Stop()
+	timeoutExceeded := time.After(timeout)
+	for {
+		select {
+		case <-timeoutExceeded:
+			return false, fmt.Errorf("db connection failed after %s timeout", timeout)
```
Comment (acekingke) [inline cmd/xenon/main.go:610]: timeout is int64, use %s is very dangerous.

File: cmd/xenon/main.go [hunk 15]
```diff
@@ -86,3 +286,366 @@ func myself(role string) MySQLNode {
...
+		case <-ticker.C:
```
Comment (acekingke) [inline cmd/xenon/main.go:611]: need times to retry. 1 second is two frequent.

**Pull request merged**